lib/active_merchant/billing/gateways/nuvei.rb

```diff
@@ -0,0 +1,255 @@
+
+        first_name, last_name = add_customer_names(address[:name], payment)
+
+        post[:billingAddress] = {
+          email: options[:email],
+          country: address[:country],
+          phone: options[:phone] || address[:phone],
+          firstName: first_name,
+          lastName: last_name
+        }.compact
+      end
+
+      def current_timestamp
+        Time.now.utc.strftime('%Y%m%d%H%M%S')
+      end
+
+      def build_post_data(post, action)
+        post[:merchantId] = @options[:merchant_id]
+        post[:merchantSiteId] = @options[:merchant_site_id]
+        post[:timeStamp] = current_timestamp.to_i
+        post[:clientRequestId] = SecureRandom.uuid
+        post[:clientUniqueId] = SecureRandom.hex(16)
+      end
+
+      def calculate_checksum(post, action)
+        case action
+        when :authenticate
+          Digest::SHA256.hexdigest("#{post[:merchantId]}#{post[:merchantSiteId]}#{post[:clientRequestId]}#{post[:timeStamp]}#{@options[:secret_key]}")
+        when :capture
+          Digest::SHA256.hexdigest("#{post[:merchantId]}#{post[:merchantSiteId]}#{post[:clientRequestId]}#{post[:clientUniqueId]}#{post[:amount]}#{post[:currency]}#{post[:relatedTransactionId]}#{post[:timeStamp]}#{@options[:secret_key]}")
+        else
+          Digest::SHA256.hexdigest("#{post[:merchantId]}#{post[:merchantSiteId]}#{post[:clientRequestId]}#{post[:amount]}#{post[:currency]}#{post[:timeStamp]}#{@options[:secret_key]}")
+        end
+      end
+
+      def send_request(post)
+        post[:checksum] = calculate_checksum(post, 'authenticate')
+
+        response = parse(ssl_post(url(:authenticate), post.to_json, headers)).with_indifferent_access
+        expiration_time = post[:timeStamp]
+        @options[:session_token] = response.dig('sessionToken')
+        @options[:token_expires] = expiration_time
+        @options[:valid_credentials] = true
+
+        Response.new(
+          response[:sessionToken].present?,
+          message_from(response),
+          response,
+          test: test?,
+          error_code: response[:errCode]
+        )
+      end
+
+      def fetch_session_token(post = {})
+        build_post_data(post, :authenticate)
+        send_request(post)
+      end
+
+      def session_token_valid?
+        return false unless @options[:session_token] && @options[:token_expires]
+
+        token_time = @options[:token_expires].to_i
+        current_time = Time.now.utc.to_i
+        (current_time - token_time) < 15 * 60
+      end
+
+      def commit(action, post, authorization = nil, method = :post)
+        MultiResponse.run do |r|
+          r.process { fetch_session_token } unless session_token_valid?
+          r.process do
+            api_request(action, post, authorization, method).tap do |response|
+              response.params.merge!(@options.slice(:session_token, :token_expires)) if @options[:valid_credentials]
+            end
+          end
+        end
```

Comment:
I think we should try to keep the established pattern for the `commit` method if we can. Typically, it handles the request, parses the response, and builds the standardized response. I wonder if the logic to refresh the session token would be better served in the `init` for the gateway class.

Reply:
yes, you're right we can add the session token request in the init, let me update it, thanks